src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,74 @@
 package mate.academy.lib;
 
+import java.lang.reflect.Field;
+import java.util.HashMap;
+import java.util.Map;
+import mate.academy.service.FileReaderService;
+import mate.academy.service.ProductParser;
+import mate.academy.service.ProductService;
+import mate.academy.service.impl.FileReaderServiceImpl;
+import mate.academy.service.impl.ProductParserImpl;
+import mate.academy.service.impl.ProductServiceImpl;
+
 public class Injector {
     private static final Injector injector = new Injector();
+    private final Map<Class<?>, Object> instances = new HashMap<>();
 
     public static Injector getInjector() {
         return injector;
     }
 
     public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+        if (!interfaceClazz.isAnnotationPresent(Component.class) && !interfaceClazz.isInterface()) {
+            throw new RuntimeException("Class " + interfaceClazz.getName()
+                    + " is not marked as @Component");
+        }
+
+        if (instances.containsKey(interfaceClazz)) {
+            return instances.get(interfaceClazz);
+        }
+
+        Object instance = createInstance(interfaceClazz);
+        instances.put(interfaceClazz, instance);
+        return instance;
+    }
+
+    private Object createInstance(Class<?> clazz) {
+        try {
+            if (clazz.isInterface()) {
+                clazz = findImplementation(clazz);
+            }
+
+            if (!clazz.isAnnotationPresent(Component.class)) {
+                throw new RuntimeException("Class " + clazz.getName()
+                        + " is not marked as @Component");
+            }
+
+            Object instance = clazz.getDeclaredConstructor().newInstance();
+
+            for (Field field : clazz.getDeclaredFields()) {
+                if (field.isAnnotationPresent(Inject.class)) {
+                    field.setAccessible(true);
+                    Object fieldInstance = getInstance(field.getType());
+                    field.set(instance, fieldInstance);
+                }
+            }
+
+            return instance;
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to create instance of class " + clazz.getName(), e);
```

Comment:
When throwing an exception, add an informative message to it. Also, don't forget to add the exception you're catching in the catch block to the RuntimeException object you throw. This will help in debugging by providing more context about the error.